torch/_functorch/aot_autograd.py

```diff
@@ -1754,12 +1772,14 @@ def functionalized_f_helper(*args):
                     # We found an input that had a (data-only) mutation.
                     # Since keep_input_mutations is set, we need to faithfully apply a copy_()
                     # so the compiler will see the input mutation in the graph.
-                    assert inpt_new is not inpt_old
-                    if meta.input_info[i].mutations_hidden_from_autograd:
-                        with torch.no_grad(), torch.autograd._unsafe_preserve_version_counter(inpt_old):
+                    if not trace_joint:
+                        assert inpt_new is not inpt_old
+                    if not trace_joint or (trace_joint and not inpt_old.requires_grad):
```

Comment:
Should test for user input that requires grad

Reply:
It is OK to test requires_grad here because the inner function is performing autograd, so it must be that we have accurate autograd information on the arguments being fed in there. (If we commuted these two wrappers, this would no longer necessarily be true)

Reply:
From a code structure perspective, I feel it would be better if this conditional was based on information in meta.input_info, rather than directly querying inpt_old, because the decision you make here affects the calling conventions everywhere else, which I assume is the plumbing you're doing...